c/tskit/tables.c

```diff
@@ -5388,6 +5381,489 @@ ancestor_mapper_run(ancestor_mapper_t *self)
     return ret;
 }
 
+/*************************
+ * IBD finder
+ *************************/
+
+static tsk_segment_t *TSK_WARN_UNUSED
+tsk_ibd_finder_alloc_segment(
+    tsk_ibd_finder_t *self, double left, double right, tsk_id_t node)
+{
+    tsk_segment_t *seg = NULL;
+
+    seg = tsk_blkalloc_get(&self->segment_heap, sizeof(*seg));
+    if (seg == NULL) {
+        goto out;
+    }
+    seg->next = NULL;
+    seg->left = left;
+    seg->right = right;
+    seg->node = node;
+
+out:
+    return seg;
+}
+
+static int TSK_WARN_UNUSED
+tsk_ibd_finder_add_output(
+    tsk_ibd_finder_t *self, double left, double right, tsk_id_t node_id, int pair_num)
+{
+    int ret = 0;
+    tsk_segment_t *tail = self->ibd_segments_tail[pair_num];
+    tsk_segment_t *x;
+
+    assert(left < right);
+    if (tail == NULL) {
+        x = tsk_ibd_finder_alloc_segment(self, left, right, node_id);
+        if (x == NULL) {
+            ret = TSK_ERR_NO_MEMORY;
+            goto out;
+        }
+        self->ibd_segments_head[pair_num] = x;
+        self->ibd_segments_tail[pair_num] = x;
+    } else {
+        x = tsk_ibd_finder_alloc_segment(self, left, right, node_id);
+        if (x == NULL) {
+            ret = TSK_ERR_NO_MEMORY;
+            goto out;
+        }
+        tail->next = x;
+        self->ibd_segments_tail[pair_num] = x;
+    }
+out:
+    return ret;
+}
+
+static int TSK_WARN_UNUSED
+tsk_ibd_finder_add_ancestry(tsk_ibd_finder_t *self, tsk_id_t input_id, double left,
+    double right, tsk_id_t output_id)
+{
+    int ret = 0;
+    tsk_segment_t *tail = self->ancestor_map_tail[input_id];
+    tsk_segment_t *x = NULL;
+
+    assert(left < right);
+    if (tail == NULL) {
+        x = tsk_ibd_finder_alloc_segment(self, left, right, output_id);
+        if (x == NULL) {
+            ret = TSK_ERR_NO_MEMORY;
+            goto out;
+        }
+        self->ancestor_map_head[input_id] = x;
+        self->ancestor_map_tail[input_id] = x;
+    } else {
+        // if (tail->right == left && tail->node == output_id) {
```

Comment:
Oh, because it leads to incorrect IBD output (you end up with the `path_ibd=False` version of IBD, which we agreed a while ago should not be the default). We want to record distinct IBD segments for intervals that have been separated by a recombination event before being passed to the samples, even if the common ancestor of both intervals is the same. I should have actually removed this instead of commenting it out, though